hyperhq/hyperd
# Implement service with IPVS

Replace service implement with IPVS.

This PR is implement through manipulating `ipvsadm` tool to admin ipvs rules. So Add/Delete/UpdateService  operation will be converted to corresponding `ipvsadm` rules, here I call them  patch.

**How to test this:**

1. build a hyperstart image with ipvs support from https://github.com/hyperhq/hyperstart/pull/247.
2. build a hyperd with this patch. (already have runv parts(https://github.com/hyperhq/runv/pull/409) vendored)
3. start hyperd with cmd:
```
hyperd -v=3 --logtostderr --host=127.0.0.1:1246
```
4. start a server pod, such as nginx. and a check pod, busybox is fine.
```
# hyperctl run -d nginx
POD id is nginx-6290370006
Time to run a POD is 1518 ms

# hyperctl run -t busybox sh    //podID2
POD id is busybox-4084967684
Time to run a POD is 1412 ms
```
5. create service rules through curl api
```
# curl -X POST "http://localhost:1246/service/add?podId=busybox-4084967684&services=\[\{\"serviceIP\":\"1.1.1.1\",\"protocol\":\"TCP\",\"servicePort\":80,\"hosts\":\[\{\"hostIP\":\"192.168.123.2\",\"hostPort\":80\}\]\}\]"
["Result=success"]

# curl -X GET "http://localhost:1246/service/list?podId=busybox-4084967684"
[{"serviceIP":"1.1.1.1","protocol":"TCP","servicePort":80,"hosts":[{"hostIP":"192.168.123.2","hostPort":80}]}]

# hyperctl exec -t busybox-4084967684 sh
//before https://github.com/hyperhq/hyperd/pull/492 merge, we need add it manually.
# route add default gw 192.168.123.1 dev eth0 
# wget 1.1.1.1
Connecting to 1.1.1.1 (1.1.1.1:80)
index.html           100% |*******************************|   612   0:00:00 ETA
```

/cc @feiskyer @resouer
